x/shield/keeper/rewards.go

```diff
@@ -60,19 +62,59 @@ func (k Keeper) PayoutNativeRewards(ctx sdk.Context, addr sdk.AccAddress) (sdk.C
 
 // DistributeFees distributes service fees to all providers.
 func (k Keeper) DistributeFees(ctx sdk.Context) {
-	serviceFees := k.GetServiceFees(ctx)
-	totalCollateral := k.GetTotalCollateral(ctx)
+	totalServiceFees := k.GetServiceFees(ctx)
+	serviceFeesLeft := k.GetServiceFeesLeft(ctx)
+	// currentBlockSecond - previousBlockSecond
+	secondsFromLastDistribution := sdk.NewDecFromInt(sdk.NewInt(int64(ctx.BlockTime().Sub(ctx.WithBlockHeight(ctx.BlockHeight() - 1).BlockTime()).Seconds())))
+	// (currentBlockSecond - previousBlockSecond) * totalServiceFees / protectionPeriodSeconds
+	serviceFees := totalServiceFees.MulDec(secondsFromLastDistribution).QuoDec(sdk.NewDecFromInt(sdk.NewInt(int64(k.GetPoolParams(ctx).ProtectionPeriod.Seconds()))))
+	bondDenom := k.BondDenom(ctx)
+	if serviceFeesLeft.Native.AmountOf(bondDenom).LT(serviceFees.Native.AmountOf(bondDenom)) {
+		serviceFees.Native = serviceFeesLeft.Native
+	}
 
+	totalCollateral := k.GetTotalCollateral(ctx)
 	providers := k.GetAllProviders(ctx)
 	for _, provider := range providers {
-		proportion := sdk.NewDecFromInt(sdk.MaxInt(provider.Collateral, sdk.ZeroInt())).QuoInt(totalCollateral)
-		nativeFees := serviceFees.Native.MulDecTruncate(proportion)
-
-		serviceFees.Native = serviceFees.Native.Sub(nativeFees)
-
+		// fees * providerCollateral / totalCollateral
+		nativeFees := serviceFees.Native.MulDec(sdk.NewDecFromInt(provider.Collateral).QuoInt(totalCollateral))
 		provider.Rewards = provider.Rewards.Add(types.MixedDecCoins{Native: nativeFees})
 		k.SetProvider(ctx, provider.Address, provider)
+
+		serviceFeesLeft.Native = serviceFeesLeft.Native.Sub(nativeFees)
 	}
+	k.SetServiceFeesLeft(ctx, serviceFeesLeft)
+	k.UpdateServiceFees(ctx)
```

Comment:
I thought this logic should be merged with deleting expiring purchases? Why two loops?